### PR TITLE
Adding ability to move between Kafka source and file source for log queries

### PR DESCRIPTION
**Pull request by poornachandra**
JIRA https://issues.cask.co/browse/CDAP-2053

### CODE REVIEW

**Comment by poornachandra** [comment]
Build - http://builds.cask.co/browse/CDAP-DUT1589-3

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/kafka/KafkaConsumer.java [hunk 1]
```diff
@@ -147,21 +133,22 @@ public long fetchOffset(Offset offset) {
...
+      // Otherwise throw exception.
+      if (timeMillis != EARLIEST_OFFSET) {
+        return fetchOffsetBefore(EARLIEST_OFFSET);
+      }
+
       String message =
         String.format("Got zero offsets in offset response for time %s from broker %s:%d for topic %s, partiton %d",
-                      offset, consumer.host(), consumer.port(), topic, partition);
+                      timeMillis, consumer.host(), consumer.port(), topic, partition);
+      closeConsumer();
```
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/kafka/KafkaConsumer.java]: Should `closeConsumer` be in a `finally` block?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/read/DistributedLogReader.java [hunk 2]
```diff
@@ -17,41 +17,88 @@
...
+    }
+
     kafkaLogReader.getLogNext(loggingContext, fromOffset, maxEvents, filter, callback);
   }
 
   @Override
   public void getLogPrev(final LoggingContext loggingContext, final LogOffset fromOffset, final int maxEvents,
                               final Filter filter, final Callback callback) {
+    // If latest logs are not requested, try reading from file.
+    if (fromOffset != LogOffset.LATEST_OFFSET) {
```
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/read/DistributedLogReader.java]: EARLIEST_OFFSET ?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/read/DistributedLogReader.java [hunk 3]
```diff
@@ -17,41 +17,88 @@
...
+    this.checkpointManager = checkpointManagerFactory.create(KafkaTopic.getTopic(),
+                                                             KafkaLogWriterPlugin.CHECKPOINT_ROW_KEY_PREFIX);
+    this.partitioner = partitioner;
   }
 
   @Override
   public void getLogNext(final LoggingContext loggingContext, final LogOffset fromOffset, final int maxEvents,
                               final Filter filter, final Callback callback) {
+    // If latest logs are not requested, try reading from file.
+    if (fromOffset != LogOffset.LATEST_OFFSET) {
```
Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/read/DistributedLogReader.java]: One problem with this logic is that we might not full fill the maxEvents requested. Maybe open a JIRA to fix this in 3.1.0

Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/read/DistributedLogReader.java]: Filed https://issues.cask.co/browse/CDAP-2246

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/kafka/KafkaConsumer.java (hunk 1 above)
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/kafka/KafkaConsumer.java]: Good point.

**Comment by poornachandra** [comment]
Will merge after build completes - http://builds.cask.co/browse/CDAP-DUT1589-7

**Pull request merged**